File: 04_BehaviouralDesignPatterns/04_CommandPattern/Main.java

import java.util.Stack;

class Light {
    public void turnOn() {
        System.out.println("Light is ON");
    }

    public void turnOff() {
        System.out.println("Light is OFF");
    }
}

class AC {
    public void turnOn() {
        System.out.println("AC is ON");
    }

    public void turnOff() {
        System.out.println("AC is OFF");
    }
}

class Fan {
    public void turnOn() {
        System.out.println("Fan is ON");
    }

    public void turnOff() {
        System.out.println("Fan is OFF");
    }
}

interface Command {
    void execute();

    void undo();
}

class TurnOnLightCommand implements Command {
    private Light light;

    public TurnOnLightCommand(Light light) {
        this.light = light;
    }

    @Override
    public void execute() {
        light.turnOn();
    }

    @Override
    public void undo() {
        light.turnOff();
    }
}

class TurnOnACCommand implements Command {
    private AC ac;

    public TurnOnACCommand(AC ac) {
        this.ac = ac;
    }

    @Override
    public void execute() {
        ac.turnOn();
    }

    @Override
    public void undo() {
        ac.turnOff();
    }
}

class TurnOnFanCommand implements Command {
    private Fan fan;

    public TurnOnFanCommand(Fan fan) {
        this.fan = fan;
    }

    @Override
    public void execute() {
        fan.turnOn();
    }

    @Override
    public void undo() {
        fan.turnOff();
    }
}

class TurnOffLightCommand implements Command {
    private Light light;

    public TurnOffLightCommand(Light light) {
        this.light = light;
    }

    @Override
    public void execute() {
        light.turnOff();
    }

    @Override
    public void undo() {
        light.turnOn();
    }
}

class TurnOffACCommand implements Command {
    private AC ac;

    public TurnOffACCommand(AC ac) {
        this.ac = ac;
    }

    @Override
    public void execute() {
        ac.turnOff();
    }

    @Override
    public void undo() {
        ac.turnOn();
    }
}

class TurnOffFanCommand implements Command {
    private Fan fan;

    public TurnOffFanCommand(Fan fan) {
        this.fan = fan;
    }

    @Override
    public void execute() {
        fan.turnOff();
    }

    @Override
    public void undo() {
        fan.turnOn();
    }
}

// Invoker
// 6 command
class RemoteControl {
    private Command[] buttons = new Command[6];
    private Stack<Command> commandHistory = new Stack<>();

    public void setCommand(int slot, Command command) {
        buttons[slot] = command;
    }

    public void pressButton(int slot) {
        if (buttons[slot] != null) {
            buttons[slot].execute();
            commandHistory.push(buttons[slot]);
        }
    }

    public void pressUndo() {
        if (!commandHistory.isEmpty()) {
            Command lastCommand = commandHistory.pop();
            lastCommand.undo();
        }
    }
}

public class Main {
    public static void main(String[] args) {
        Light light = new Light();
        AC ac = new AC();
        Fan fan = new Fan();

        RemoteControl remote = new RemoteControl();

        remote.setCommand(0, new TurnOnLightCommand(light));
        remote.setCommand(1, new TurnOffLightCommand(light));
        remote.setCommand(2, new TurnOnACCommand(ac));
        remote.setCommand(3, new TurnOffACCommand(ac));
        remote.setCommand(4, new TurnOnFanCommand(fan));
        remote.setCommand(5, new TurnOffFanCommand(fan));

        remote.pressButton(0); // Light ON
        remote.pressButton(2); // AC ON
        remote.pressButton(4); // Fan ON

        remote.pressUndo(); // Fan OFF
        remote.pressUndo(); // AC OFF
        remote.pressUndo(); // Light OFF
    }
}

File: 04_BehaviouralDesignPatterns/04_CommandPattern/cpp/Main.cpp
#include <iostream>
#include <string>
#include <memory>
#include <stack>

// Receivers
class Light {
public:
    void turnOn() {
        std::cout << "Light is ON" << std::endl;
    }

    void turnOff() {
        std::cout << "Light is OFF" << std::endl;
    }
};

class AC {
public:
    void turnOn() {
        std::cout << "AC is ON" << std::endl;
    }

    void turnOff() {
        std::cout << "AC is OFF" << std::endl;
    }
};

class Fan {
public:
    void turnOn() {
        std::cout << "Fan is ON" << std::endl;
    }

    void turnOff() {
        std::cout << "Fan is OFF" << std::endl;
    }
};

// Command interface
class Command {
public:
    virtual void execute() = 0;
    virtual void undo() = 0;
    virtual ~Command() = default;
};

// Concrete Commands for Light
class LightOnCommand : public Command {
private:
    Light& light;

public:
    LightOnCommand(Light& light) : light(light) {}

    void execute() override {
        light.turnOn();
    }

    void undo() override {
        light.turnOff();
    }
};

class LightOffCommand : public Command {
private:
    Light& light;

public:
    LightOffCommand(Light& light) : light(light) {}

    void execute() override {
        light.turnOff();
    }

    void undo() override {
        light.turnOn();
    }
};

// Concrete Commands for AC
class ACOnCommand : public Command {
private:
    AC& ac;

public:
    ACOnCommand(AC& ac) : ac(ac) {}

    void execute() override {
        ac.turnOn();
    }

    void undo() override {
        ac.turnOff();
    }
};

class ACOffCommand : public Command {
private:
    AC& ac;

public:
    ACOffCommand(AC& ac) : ac(ac) {}

    void execute() override {
        ac.turnOff();
    }

    void undo() override {
        ac.turnOn();
    }
};

// Concrete Commands for Fan
class FanOnCommand : public Command {
private:
    Fan& fan;

public:
    FanOnCommand(Fan& fan) : fan(fan) {}

    void execute() override {
        fan.turnOn();
    }

    void undo() override {
        fan.turnOff();
    }
};

class FanOffCommand : public Command {
private:
    Fan& fan;

public:
    FanOffCommand(Fan& fan) : fan(fan) {}

    void execute() override {
        fan.turnOff();
    }

    void undo() override {
        fan.turnOn();
    }
};

// Invoker
class RemoteControl {
private:
    std::stack<std::shared_ptr<Command>> history;

public:
    void pressButton(std::shared_ptr<Command> command) {
        command->execute();
        history.push(command);
    }

    void pressUndo() {
        if (!history.empty()) {
            auto command = history.top();
            command->undo();
            history.pop();
        }
    }
};

int main() {
    // Create receivers
    Light light;
    AC ac;
    Fan fan;

    // Create commands
    auto lightOn = std::make_shared<LightOnCommand>(light);
    auto lightOff = std::make_shared<LightOffCommand>(light);
    auto acOn = std::make_shared<ACOnCommand>(ac);
    auto acOff = std::make_shared<ACOffCommand>(ac);
    auto fanOn = std::make_shared<FanOnCommand>(fan);
    auto fanOff = std::make_shared<FanOffCommand>(fan);

    // Create invoker
    RemoteControl remote;

    // Use the remote
    remote.pressButton(lightOn);
    remote.pressButton(acOn);
    remote.pressButton(fanOn);
    remote.pressUndo(); // Undo fan on
    remote.pressButton(lightOff);
    remote.pressUndo(); // Undo light off

    return 0;
}